src/alarms/alarm-handling.md

```diff
@@ -0,0 +1,87 @@
+
+A lost alam-raise could be even more problematic if the raised alarm requires some cloud-site interaction (or manual interaction). Since the lost alarm is not visible on cloud no intercation will be started. And since not state-change happen again, the alarm will never appear at cloud.
+
+**Consequence:**<br/>
+All alarm state-changes need to be transfered reliable from **Application to thin-edge**, and from **thin-edge to cloud**.
+I.E. alarm state-changes must not got lost, or device software must be able to detect loss and react accordingly (e.g. retry to transfer).
+
+More details about loss off data-transfer and proper reaction see...  TODO: add data-flow-image here!
+
+# Information Set per Alarm
+The specified alarm attributes below are inspired by cumulocity data model for alarms, but shall be re-usable for other clouds where possible.
+
+| Name              | Description   |
+| ----------------- |---------------|
+| ref-string        | Device-unique reference string (kind of alarm ID or alarm name), used to reference the once occured alarm again, e.g. in case of updates (e.g. "temperature_sensor_loss") |
+| message-string    | Human readable short information about alarm reason (e.g. "Temperature sensor does not respond") |
+| severity-string   | Could be "CRITICAL", "MAJOR", "MINOR" or "WARNING" |
+| status-string     | Could be "ACTIVE" or "CLEARED"<br/>
+| time-string       | Timestamp indicating when the alarm (or the alarm-update) had occured (in [ISO 8601 format](http://www.w3.org/TR/NOTE-datetime#)) |
+
+# Data flow
+
+Figure below illustrates the data flow from Customer Application broker/thin-edge up to cloud. It shows especially the use of MQTT QOS=1, MQTT Retain and Mosquitto Persistence to achieve reliability alarm state-change transfer.
+
+![Sequence Diagram Update SW-list](images/alarm_dataflow.svg)
+
+**TODO:** Interface from Mapper to Cloud to be defined. Two options are possible:
+1) **JSONviaMQTT:** To have reliability use QOS=1. Open issue: Not yet completely confirmed that C8Y sends PUBACK when msg was processed, instead just on arrival.
+2) **HTTP REST:** HTTP response indicates if message was processed sucessfully. Anyway HTTP REST would increase complexity of implementation and increase data traffic.
+
+
+# Public MQTT-based alarm interface
+
+Similar to thin-edge'S measurement interface, alarm interface is based on MQTT topics.
+
+## Topic structure and payload
+
+Proposal to have "severity" and "ref-string" as topics:
+
+**thin-edge JSON (alarm) format:**
+```
+Topic:   tedge/alarms/<device-id>/<severity-string>/<ref-string>
+Payload: {
+  "message":  <message-string>,
+  "status":   <status-string>,  

```

Comment:
As we are utilizing the topics for the most of the handling why not push out the status there as well. It gives you even more flexibility with topics and you can only subscribe to ACTIVEation of the alarm.

Reply:
Since we are using a retain flag, one can even imply the status. I mean the status is encoded neither in the topic  nor in the message, but by the fact there is a retained message or not. Only one thing is no more representable: the datetime since then an alarm is off.

Reply:
Representing CLEARing of an alarm with an empty message would be a useful addition simplifying the contract for the end-user by not forcing him to provide a `message` or even a `timestamp` for clearing a message. Fortunately, the "raw payload" support enables just that.